### PR TITLE
Plugin based on chosen for dropdowns

### PR DESCRIPTION
**Pull request by premithk**
**Merge request checklist**

- [x] I read the [guidelines for contributing](https://github.com/mistic100/jQuery-QueryBuilder/blob/master/.github/CONTRIBUTING.md)
- [x] I created my branch from `dev` and I am issuing the PR to `dev`
- [x] I didn't pushed the `dist` directory
- [x] Unit tests are OK
- [x] If it's a new feature, I added the necessary unit tests
- [ ] If it's a new language, I filled the `__locale` and `__author` fields

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -17,7 +17,8 @@
     "bootstrap": ">=3.1.0 <4",
     "moment": ">=2.6.0",
     "jquery-extendext": ">=0.1.2",
-    "doT": ">=1.0.3"
+    "doT": ">=1.0.3",
+    "chosen": "^1.8.3"
```
Comment (mistic100) [inline bower.json]: it cannot be an explicit dependency, move this to examples/bower.json

File: src/plugins/chosen-selectpicker/plugin.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    this.on('afterUpdateRuleFilter', function(e, rule) {
+        rule.$el.find(Selectors.rule_filter).trigger('chosen:updated');
+    });
+
+    this.on('afterUpdateRuleOperator', function(e, rule) {
+        rule.$el.find(Selectors.rule_operator).trigger('chosen:updated');
+    });
+
+    this.on('beforeDeleteRule', function(e, rule) {
+        if ($('.form-control').length !== 0 && $('.form-control').prop('nodeName') === 'SELECT') {
```
Comment (mistic100) [inline src/plugins/chosen-selectpicker/plugin.js]: This test seems totally wrong :

- You are querying `.form-control` in the whole page, not in the current rule only
- I don't understand what is the goal as you remove the `form-control` class on lines 22 and 26, the bt-selectpicker plugin does not have this test and works perfectly

File: bower.json (hunk 1 above)
Comment (premithk) [reply bower.json]: Yes. Have moved this to examples/bower.json

File: src/plugins/chosen-selectpicker/plugin.js (hunk 2 above)
Comment (premithk) [reply src/plugins/chosen-selectpicker/plugin.js]: Yes @mistic100 totally unwanted check. was a left over code from something i was trying before you had given me clue on how to check for bt-select. have removed it. thank you

File: bower.json [hunk 3]
```diff
@@ -17,7 +17,7 @@
     "bootstrap": ">=3.1.0 <4",
     "moment": ">=2.6.0",
     "jquery-extendext": ">=0.1.2",
-    "doT": ">=1.0.3"
+    "doT": ">=1.0.3",
```
Comment (mistic100) [inline bower.json]: the JSON is now invalid (check TravisCI)

Comment (premithk) [reply bower.json]: yes, a trailing comma was there, fixed it

File: examples/bower.json [hunk 4]
```diff
@@ -1,8 +1,9 @@
 {
   "name": "jQuery-QueryBuilder-example",
-  "dependencies" : {
+  "dependencies": {
     "seiyria-bootstrap-slider": "latest",
     "bootswatch-dist": "#slate",
-    "selectize": "latest"
+    "selectize": "latest",
+    "chosen": "^1.8.3",
```
[the remark this answers is not in the document]
Comment (premithk) [reply examples/bower.json]: ah. changed it

File: src/plugins/chosen-selectpicker/plugin.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    if (this.settings.plugins['bt-selectpicker']) {
+        Utils.error('Conflict', 'bt-selectpicker is already selected as the dropdown plugin. Please remove chosen-selectpicker from the plugin list');
+    }
+
+    var Selectors = QueryBuilder.selectors;
+
+    // init selectpicker
+    this.on('afterCreateRuleFilters', function(e, rule) {
+        rule.$el.find(Selectors.rule_filter).removeClass('form-control').chosen();
```
Comment (mistic100) [inline src/plugins/chosen-selectpicker/plugin.js]: please actually pass the `options` to chosen (same on line 26)

last change before merge

Comment (premithk) [reply src/plugins/chosen-selectpicker/plugin.js]: oh yes. have changed. thank you

**Comment by mistic100** [comment]
Thank you for your time

**Comment by premithk** [comment]
Thank you. One more thing, should i add a check in bt-selectpicker to check if chosen-picker is already applied? If yes shall i open a pull request some time next week?

**Comment by mistic100** [comment]
yeah good idea

**Pull request merged**